Replace element position with brackets

tonysun83 opened the pull request:
To be consistent with view based indexes, we allow the user
to access array elements via position. We convert that position
to [] for the underlying indexed field. For example, myarray.0.1,
would be converted to myarray.[].[]. This converted field will then
be used for the text search. The results will then be filtered by
the match function.

Fogbugzid: 46268

davisp commented on src/mango_selector_text.erl:
```diff
@@ -159,9 +159,17 @@ convert(Path, {[{<<"$size">>, Arg}]}) ->
+        _ ->
+            re:replace(Field0,<<"\\.\\d+">>,<<"\\.\\[\\]">>,[global,
+                {return,binary}])
+    end,
```
I'm disliking that there are three uncompiled regexps when we have to make this change. And the old single regexp version looks broken anyway.

https://github.com/cloudant/mango/blob/46268-fix-match-arrays/src/mango_doc.erl#L540-L564

I'd move those mango_doc functions to mango_util, then parse the field once here adding a conversion function something like:

    convert(Path, {[{Field0, Cond}]}) ->
        {ok, PathParts0} = mango_util:parse_field(Field0),
        PathParts1 = replace_array_indexes(Path0),
        NewPath = lists:reverse(PathParts1) ++ Path,
        convert(NewPath, Cond);


    replace_array_indexes([]) ->
        [];
    replace_array_indexes([Part | Rest]) ->
        NewParts = try
            list_to_integer(binary_to_list(Part)),
            <<"[]">>;
        catch _:_ ->
            Part
        end,
        [NewPart | replace_array_indexes(Rest)].

tonysun83 commented:
Changed with your suggestions. I had to add a replacement to accommodate using parse_field. The problem is that we are slightly constrained by our escaping mechanism. I didn't want to go re-vamping the whole thing which might break other parts of our code.

Old way that worked given the order of our path array:
1) Use re:split("\\."). <<"a\\."b">> was split to
 [<<"a\\">>, <<"b">>]
2) Later on, we join these two binaries and get <<"a\\.b">>
3) We call parse_field which converts this to <<"a.b">>
4) We call lucene_escape_field which "correctly" translates to <<"a_2eb">>

New way that calls parse_field twice:
1) Call parse_field to convert to parts. <<"a\\."b">> is converted to [<<"a.b">>] 
2) Replace . with \\. to get  [<<"a\\.b">>]
3) We call parse_field which converts this to <<"a.b">>
4) We call lucene_escape_field which "correctly" translates to <<"a_2eb">>

davisp commented:
I'm suddenly wondering what happens to this code if someone tries to index a document like

    {"foo": {"1": "some stuff"}}

Ie, the field is a string but looks like a number.

davisp commented:
Ie, do we need to be adding an OR query in here to search for both? IIRC the mango_doc:get_field call checks for both versions.

tonysun83 commented:
We'll need to add something for numbers. Right now, I think the code would change that to ".".

tonysun83 commented:
Added the additional OR query to handle numbers. However, I'm not sure how we can handle nested numbered fields with arrays. For instance "1" : [{"2" : "nested"}]. "1.0.2" : "nested" would be the query, but I don't think we could every satisfy that. Maybe a limitation we have to live with?

davisp commented on src/mango_selector_text.erl:
```diff
@@ -159,12 +159,35 @@ convert(Path, {[{<<"$size">>, Arg}]}) ->
+    NewPath = PP2 ++ Path,
+    case HasInteger of
+        true ->
+            OldPath = lists:reverse(PP1) ++ Path,
```
You want lists:reverse(PP1, Path) here.

davisp commented on src/mango_selector_text.erl:
```diff
@@ -159,12 +159,35 @@ convert(Path, {[{<<"$size">>, Arg}]}) ->
+    % an incorrect query since we need [<<"a.b">>]. Without breaking
+    % our escaping mechanism, we simply revert this first parse_field
+    % effect and replace instances of "." to "\\.".
+    PP1 = [re:replace(P, <<"\\.">>, <<"\\\\.">>,
```
We should precompile this regexp like you did for detecting doubles.

davisp commented:
For your comments, the trick that we have to consider is that a number is either an array index or a string and we have to check for both. I'm also not certain how we couldn't handle your "1.0.2" example above. Near as I can tell it should just work unless I'm forgetting something.

+1 to squerge once you address the last two minor comments I had.
